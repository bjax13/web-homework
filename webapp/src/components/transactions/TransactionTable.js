import React, { useState, useEffect } from 'react'
import { arrayOf, string, bool, number, shape } from 'prop-types'
import { css } from '@emotion/core'
import { useMutation } from '@apollo/client'
import { FormattedMessage, useIntl } from 'react-intl'
import CreateTransaction from '../../gql/transactionsCreate.gql'
import UpdateTransaction from '../../gql/transactionsUpdate.gql'
import DeleteTransaction from '../../gql/transactionsDelete.gql'
import TransactionTableRow from './TransactionTableRow/TransactionTableRow'
import { getRowCells, TRANSACTION_TYPES, columnProperties } from './TransactionTable.utils'
import CreateEditTransactionModal from './createEditTransactionModal/CreateEditTransactionModal'
import TransactionTableColumnHeader from './TransactionTableColumnHeader/TransactionTableColumnHeader'
import Button from '../general/Button/Button'
import TransactionChart from './transactionChart/TransactionChart'
import messages from './TransactionTable.messages'

const styles = css`
  .header {
    font-weight: bold;
  }
`
function TransactionTable ({ data }) {
  const [createTransaction] = useMutation(CreateTransaction)
  const [updateTransaction] = useMutation(UpdateTransaction)
  const [deleteTransaction] = useMutation(DeleteTransaction)
  const [tabledata, setTableData] = useState(data)
  const [isModalOpen, setIsModalOpen] = useState(false)
  const [isRomanNumeral, setIsRomanNumeral] = useState(false)
  const [isModalEditMode, setModalEditMode] = useState(false)
  const [seedTransactionToEdit, setSeedTransactionToEdit] = useState({})
  const intl = useIntl()

  useEffect(() => {
    setTableData(data)
  }, [data])

  const handleEditTransaction = (transaction) => {
    setSeedTransactionToEdit(transaction)
    setModalEditMode(true)
    setIsModalOpen(true)
  }

  const handleDeleteTransaction = (id) => {
    deleteTransaction({ variables: { deleteTransactionId: id } }).then(
      ({
        data: {
          deleteTransaction: { id: deletedTransactionId }
        }
      }) => {
        const deletedTransactionIndex = tabledata.findIndex((transaction) => transaction.id === deletedTransactionId)

        if (deletedTransactionIndex !== -1) {
          const newTabledata = [...tabledata]
          newTabledata.splice(deletedTransactionIndex, 1)
          setTableData(newTabledata)
        }
      }
    )
  }

  const handleModalSubmit = (transactionValues, isEdit) => {
    const { id, amount = 0, description = '', transactionType = 'TRANSACTION_TYPES.DEBIT' } = transactionValues
    let variables = {
      amount: parseFloat(amount),
      description: description
    }

    if (transactionType === TRANSACTION_TYPES.CREDIT || TRANSACTION_TYPES.DEBIT) {
      variables = {
        ...variables,
        [transactionType]: true
      }
    }
    if (isEdit) {
      variables = {
        ...variables,
        updateTransactionId: id
      }
      updateTransaction({ variables }).then(({ data: { updateTransaction } }) => {
        const { id: updateTransactionId } = updateTransaction

        variables = {}
        setIsModalOpen(false)
        setModalEditMode(false)
        const editedTransactionIndex = tabledata.findIndex((transaction) => transaction.id === updateTransactionId)

        if (editedTransactionIndex !== -1) {
          const newTabledata = [...tabledata]
          newTabledata.splice(editedTransactionIndex, 1, updateTransaction)
          setTableData(newTabledata)
        }
      })
    } else {
      createTransaction({ variables }).then(({ data: { addTransaction } }) => {
        variables = {}
        setIsModalOpen(false)
        setTableData([...tabledata, addTransaction])
      })
    }
  }

  const handleCreateModalOpen = () => {
    setModalEditMode(false)
    setIsModalOpen(true)
  }

  const getChartData = () => {
    const chartData = tabledata.map(({ description, amount }) => [description, amount])
    chartData.unshift(['description', 'amount']) // add column header nessisary for data
    return chartData
  }

  return (
    <>
      <Button onClick={() => setIsRomanNumeral(!isRomanNumeral)}>
        <FormattedMessage {...messages.romanNumeral} />
      </Button>
      {!isModalOpen && (
        <Button onClick={handleCreateModalOpen}>
          <FormattedMessage {...messages.createTransaction} />
        </Button>
      )}
      <TransactionChart data={getChartData()} title={intl.formatMessage({ ...messages.comparitiveTransaction })} />
      <table css={styles}>
        <tbody>
          <tr className='header'>
            {columnProperties.map(({ text, type }) => (
              <TransactionTableColumnHeader key={type}>{text}</TransactionTableColumnHeader>
            ))}
          </tr>
          {tabledata.map((transaction) => (
            <TransactionTableRow
              isRomanNumeral={isRomanNumeral}
              key={transaction.id}
              onDelete={handleDeleteTransaction}
              onEdit={handleEditTransaction}
              rowCells={getRowCells(transaction)}
              transaction={transaction}
            />
          ))}
        </tbody>
      </table>
      <CreateEditTransactionModal
        isEdit={isModalEditMode}
        isOpen={isModalOpen}
        onCancel={() => setIsModalOpen(false)}
        onSubmit={handleModalSubmit}
        seedTransactionToEdit={seedTransactionToEdit}
      />
    </>
  )
}

TransactionTable.propTypes = {
  data: arrayOf(
    shape({
      id: string,
      user_id: string,
      description: string,
      merchant_id: string,
      debit: bool,
      credit: bool,
      amount: number
    })
  )
}

export default TransactionTable
