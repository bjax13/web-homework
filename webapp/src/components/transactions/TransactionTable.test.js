import { render, getByTestId, fireEvent, getByText } from '@testing-library/react'
import React from 'react'
import TransactionTable from './TransactionTable'
import { MockedProvider } from '@apollo/react-testing'
import { client } from '../../network/apollo-client'
import { IntlProvider } from 'react-intl'

const mockData = [
  {
    id: '61add94066be4c252121c11b',
    user_id: null,
    description: 'Jeffery buying Pizza',
    merchant_id: null,
    debit: true,
    credit: false,
    amount: 67
  },
  {
    id: '61add94066be4c252121c11c',
    user_id: null,
    description: 'Madeline buying blue',
    merchant_id: null,
    debit: false,
    credit: true,
    amount: 949
  },
  {
    id: '61add94066be4c252121c140',
    user_id: null,
    description: 'Cleora buying withdrawal',
    merchant_id: null,
    debit: false,
    credit: true,
    amount: 235
  }
]

describe('Transactions Table', () => {
  test('renders table', async () => {
    const { container } = render(
      <MockedProvider client={client}>
        <IntlProvider locale='en'>
          <TransactionTable data={mockData} />
        </IntlProvider>
      </MockedProvider>
    )

    expect(container).toMatchSnapshot()
  })
  test('should have all 3 transactions in the table', async () => {
    const { container } = render(
      <MockedProvider client={client}>
        <IntlProvider locale='en'>
          <TransactionTable data={mockData} />
        </IntlProvider>
      </MockedProvider>
    )

    expect(getByTestId(container, 'transaction-61add94066be4c252121c11b-id')).toHaveTextContent('61add94066be4c252121c11b')
    expect(getByTestId(container, 'transaction-61add94066be4c252121c11c-id')).toHaveTextContent('61add94066be4c252121c11c')
    expect(getByTestId(container, 'transaction-61add94066be4c252121c140-id')).toHaveTextContent('61add94066be4c252121c140')
  })
  test('nubers should be roman after toggle to roman is clicked', async () => {
    const { container } = render(
      <MockedProvider client={client}>
        <IntlProvider locale='en'>
          <TransactionTable data={mockData} />
        </IntlProvider>
      </MockedProvider>
    )

    expect(getByTestId(container, 'transaction-61add94066be4c252121c11b-amount')).toHaveTextContent('67')
    fireEvent.click(getByText(container, 'Toggle Roman Numeral'))
    expect(getByTestId(container, 'transaction-61add94066be4c252121c11b-amount')).toHaveTextContent('IXII')
  })
})
