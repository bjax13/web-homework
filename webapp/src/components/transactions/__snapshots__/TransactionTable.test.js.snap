// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Table renders table 1`] = `
.emotion-0 {
  border: 0;
  background: #78f89f;
  border-radius: 5px;
  padding: 0.5rem 1rem;
  margin: 0.2rem;
  font-size: 0.8rem;
  line-height: 1;
}

.emotion-17 .header {
  font-weight: bold;
}

.emotion-18 .modal-overlay {
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  z-index: 2;
  cursor: pointer;
}

.emotion-18 .modal {
  width: 500px;
  background: white;
  border: 1px solid #ccc;
  position: absolute;
  top: 20%;
  left: 20%;
  z-index: 5;
}

.emotion-18 .modal h2 {
  border-bottom: 1px solid #ccc;
  padding: 1rem;
  margin: 0;
}

.emotion-18 .modal .modal-content {
  padding: 1rem;
}

.emotion-18 .modal .actions {
  border-top: 1px solid #ccc;
  background: #eee;
  padding: 0.5rem 1rem;
}

.emotion-18 .modal .actions button:first-child {
  margin-right: 15px;
}

<div>
  <button
    class="emotion-0 emotion-1"
  >
    Toggle Roman Numeral
  </button>
  <button
    class="emotion-0 emotion-1"
  >
    Create Transaction
  </button>
  <div
    class="emotion-4"
  >
    <div>
      Loading ...
    </div>
  </div>
  <table
    class="emotion-17"
  >
    <tbody>
      <tr
        class="header"
      >
        <td>
          ID
        </td>
        <td>
          Description
        </td>
        <td>
          Debit
        </td>
        <td>
          Credit
        </td>
        <td>
          Amount
        </td>
        <td>
          Edit
        </td>
        <td>
          Delete
        </td>
      </tr>
      <tr
        data-testid="transaction-61add94066be4c252121c11b"
      >
        <td
          data-testid="transaction-61add94066be4c252121c11b-id"
        >
          61add94066be4c252121c11b
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11b-description"
        >
          Jeffery buying Pizza
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11b-debit"
        >
          Debit
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11b-credit"
        />
        <td
          data-testid="transaction-61add94066be4c252121c11b-amount"
        >
          67
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11b-onEdit"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Edit
          </button>
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11b-onDelete"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        data-testid="transaction-61add94066be4c252121c11c"
      >
        <td
          data-testid="transaction-61add94066be4c252121c11c-id"
        >
          61add94066be4c252121c11c
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11c-description"
        >
          Madeline buying blue
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11c-debit"
        />
        <td
          data-testid="transaction-61add94066be4c252121c11c-credit"
        >
          Credit
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11c-amount"
        >
          949
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11c-onEdit"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Edit
          </button>
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c11c-onDelete"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        data-testid="transaction-61add94066be4c252121c140"
      >
        <td
          data-testid="transaction-61add94066be4c252121c140-id"
        >
          61add94066be4c252121c140
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c140-description"
        >
          Cleora buying withdrawal
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c140-debit"
        />
        <td
          data-testid="transaction-61add94066be4c252121c140-credit"
        >
          Credit
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c140-amount"
        >
          235
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c140-onEdit"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Edit
          </button>
        </td>
        <td
          data-testid="transaction-61add94066be4c252121c140-onDelete"
        >
          <button
            class="emotion-0 emotion-1"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="emotion-18"
  />
</div>
`;
